net-misc/dahdi/dahdi-3.1.0.ebuild

```diff
@@ -0,0 +1,94 @@
+HOMEPAGE="https://www.asterisk.org"
+SRC_URI="https://downloads.asterisk.org/pub/telephony/dahdi-linux/releases/${MY_P}.tar.gz
+	https://www.junghanns.net/downloads/jnet-dahdi-drivers-${JNET}.tar.gz
+	https://downloads.uls.co.za/gentoo/dahdi/gentoo-dahdi-patchset-${GENTOO}.tar.bz2
+	https://downloads.digium.com/pub/telephony/firmware/releases/dahdi-fwload-vpmadt032-1.25.0.tar.gz
+	https://downloads.digium.com/pub/telephony/firmware/releases/dahdi-fw-a4a-a0017.tar.gz
+	https://downloads.digium.com/pub/telephony/firmware/releases/dahdi-fw-a4b-d001e.tar.gz
+	https://downloads.digium.com/pub/telephony/firmware/releases/dahdi-fw-a8a-1d0017.tar.gz
+	https://downloads.digium.com/pub/telephony/firmware/releases/dahdi-fw-a8b-1f001e.tar.gz
+	https://downloads.digium.com/pub/telephony/firmware/releases/dahdi-fw-hx8-2.06.tar.gz
+	https://downloads.digium.com/pub/telephony/firmware/releases/dahdi-fw-oct6114-032-1.05.01.tar.gz
+	https://downloads.digium.com/pub/telephony/firmware/releases/dahdi-fw-oct6114-064-1.05.01.tar.gz
+	https://downloads.digium.com/pub/telephony/firmware/releases/dahdi-fw-oct6114-128-1.05.01.tar.gz
+	https://downloads.digium.com/pub/telephony/firmware/releases/dahdi-fw-oct6114-256-1.05.01.tar.gz
+	https://downloads.digium.com/pub/telephony/firmware/releases/dahdi-fw-tc400m-MR6.12.tar.gz
+	https://downloads.digium.com/pub/telephony/firmware/releases/dahdi-fw-te133-7a001e.tar.gz
+	https://downloads.digium.com/pub/telephony/firmware/releases/dahdi-fw-te134-780017.tar.gz
+	https://downloads.digium.com/pub/telephony/firmware/releases/dahdi-fw-te435-13001e.tar.gz
+	https://downloads.digium.com/pub/telephony/firmware/releases/dahdi-fw-te436-10017.tar.gz
+	https://downloads.digium.com/pub/telephony/firmware/releases/dahdi-fw-te820-1.76.tar.gz
+	https://downloads.digium.com/pub/telephony/firmware/releases/dahdi-fw-vpmoct032-1.12.0.tar.gz
+"
+
+LICENSE="LGPL-2.1"
+SLOT="0"
+KEYWORDS="~amd64 ~ppc ~ppc64 ~x86"
+IUSE="flash oslec"
+
+DEPEND=""
+RDEPEND=""
+
+EPATCH_SUFFIX="diff"
+PATCHES=( "${WORKDIR}/dahdi-patchset" )
+
+CONFIG_CHECK="MODULES ~CRC_CCITT"
+
+pkg_pretend() {
+	use oslec && CONFIG_CHECK+=" ECHO"
+	check_extra_config
+}
+
+src_unpack() {
+	unpack ${A}
+	# Copy the firmware tarballs over, the makefile will try and download them otherwise
+	for file in ${A} ; do
+		[[ "${file}" = dahdi-fw* ]] || continue
+		cp "${DISTDIR}"/${file} "${MY_P}"/drivers/dahdi/firmware/ || die "Error copying ${file} to ${S}/${MY_P}/drivers/dahdi/firmware/"
+	done
+	# But without the .bin's it'll still fall over and die, so copy those too.
+	mv *.bin "${MY_P}"/drivers/dahdi/firmware/ || die "Error moving firmware files into the right folders."
+
+	for drv in ${JNET_DRIVERS}; do
+		ln "${WORKDIR}/jnet-dahdi-drivers-${JNET}/${drv}/${drv}.c" "${MY_P}/drivers/dahdi/" || die "Error copying ${drv}.c from jnet to DAHDI."
+		ln "${WORKDIR}/jnet-dahdi-drivers-${JNET}/${drv}/${drv}.h" "${MY_P}/drivers/dahdi/" || die "Error copying ${drv}.h from jnet to DAHDI."
```

Comment:
Copying or linking ;)

Reply:
Cosmetic to be honest.  Updated the wording on the die message.

Could we please arrange for me to get your kernel config?

Looking at (linux 5.3.11) Makefile, the lines in it's Makefile that's causing the issue:

 900 
 901 include scripts/Makefile.kasan
 902 include scripts/Makefile.extrawarn
 903 include scripts/Makefile.ubsan
 904 

Doesn't look like there is anything conditional about that.  (I did also scan up/down to see if it's not just poor indentation on nesting).  I've checked the sources for 5.4 and it looks the same ... so honestly, I don't know about that one.  Is your source tree still in tact?  (ie, could you please verify that in /usr/src/linux-5.4.0-pf2 there exists a file scripts/Makefile.ubsan?  Please also verify that /usr/src/linux symlink points correctly.

KSRC=/usr/src/linux <-- this comes from ${KV_OUT_DIR}, which actually comes from linux-info, not linux-mod (the latter inherits the former, so the question is, do I need to explicitly inherit both, or is it fine to just inherit the latter?)

I've downloaded the 5.4.8 kernel sources, and extracted, the above is just offset by one line in the primary Makefile, and the scripts/Makefile.ubsan does exist.  I'm failing to see what caused your failure.

Reply:
All of the fetches for me for dahdi 3.1.0 is looking fine.

If you could be more specific as to which downloads are failing I can see if it's possible to repair.  However, anything < 3.1.0 won't compile against anything but an insanely old kernel anyway so I suggest we focus on that.

Reply:
https://dev.gentoo.org/~juippis/logs/kernel_config_5.4.0-pf3.txt

```
$ ls -l linux
lrwxrwxrwx 1 root root 16 Jan  1 18:43 linux -> linux-5.4_p3-pf/

$ ls linux
arch     fs       lib                      modules.order   sound       vmlinux
block    include  Makefile                 Module.symvers  System.map  vmlinux.o
certs    init     mm                       net             tools
crypto   ipc      modules.builtin          scripts         usr
drivers  kernel   modules.builtin.modinfo  security        virt

$ ls linux/scripts/
asn1_compiler  conmakehash   genksyms  kconfig  pnmtologo  sign-file
basic          extract-cert  kallsyms  mod      selinux    sortextable
```

Indeed,
```
$ ls linux-5.4.8-gentoo/scripts/
adjust_autoksyms.sh                ld-version.sh
asn1_compiler.c                    leaking_addresses.pl
atomic                             Lindent
basic                              link-vmlinux.sh
bin2c.c                            Makefile
bloat-o-meter                      Makefile.asm-generic
bootgraph.pl                       Makefile.build
bpf_helpers_doc.py                 Makefile.clean
cc-can-link.sh                     Makefile.dtbinst
check_extable.sh                   Makefile.extrawarn
checkincludes.pl                   Makefile.gcc-plugins
checkkconfigsymbols.py             Makefile.headersinst
checkpatch.pl                      Makefile.host
checkstack.pl                      Makefile.kasan
checksyscalls.sh                   Makefile.kcov
checkversion.pl                    Makefile.lib
clang-version.sh                   Makefile.modbuiltin
cleanfile                          Makefile.modfinal
cleanpatch                         Makefile.modinst
coccicheck                         Makefile.modpost
coccinelle                         Makefile.modsign
config                             Makefile.package
conmakehash.c                      Makefile.ubsan
const_structs.checkpatch           makelst
decodecode                         markup_oops.pl
decode_stacktrace.sh               mkcompile_h
depmod.sh                          mkmakefile
diffconfig                         mksysmap
documentation-file-ref-check       mkuboot.sh
dtc                                mod
export_report.pl                   module-common.lds
extract-cert.c                     modules-check.sh
extract-ikconfig                   namespace.pl
extract-module-sig.pl              nsdeps
extract-sys-certs.pl               objdiff
extract-vmlinux                    package
extract_xc3028.pl                  parse-maintainers.pl
faddr2line                         patch-kernel
file-size.sh                       pnmtologo.c
find-unused-docs.sh                profile2linkerlist.pl
gcc-goto.sh                        prune-kernel
gcc-ld                             recordmcount.c
gcc-plugins                        recordmcount.h
gcc-plugin.sh                      recordmcount.pl
gcc-version.sh                     selinux
gcc-x86_32-has-stack-protector.sh  setlocalversion
gcc-x86_64-has-stack-protector.sh  show_delta
gdb                                sign-file.c
gen_compile_commands.py            sortextable.c
gen_ksymdeps.sh                    sortextable.h
genksyms                           spdxcheck.py
get_abi.pl                         spdxcheck-test.sh
get_dvb_firmware                   spelling.txt
get_maintainer.pl                  sphinx-pre-install
gfp-translate                      split-man.pl
headerdep.pl                       stackdelta
headers_check.pl                   stackusage
headers_install.sh                 subarch.include
insert-sys-cert.c                  tags.sh
kallsyms.c                         tools-support-relr.sh
Kbuild.include                     tracing
kconfig                            unifdef.c
Kconfig.include                    ver_linux
kernel-doc                         xen-hypercalls.sh
ksymoops                           xz_wrap.sh
```

tons of more stuff.


However, 
```
$ ls linux-5.4_p4-pf/scripts/
adjust_autoksyms.sh                ld-version.sh
asn1_compiler.c                    leaking_addresses.pl
atomic                             Lindent
basic                              link-vmlinux.sh
bin2c.c                            Makefile
bloat-o-meter                      Makefile.asm-generic
bootgraph.pl                       Makefile.build
bpf_helpers_doc.py                 Makefile.clean
cc-can-link.sh                     Makefile.dtbinst
check_extable.sh                   Makefile.extrawarn
checkincludes.pl                   Makefile.gcc-plugins
checkkconfigsymbols.py             Makefile.headersinst
checkpatch.pl                      Makefile.host
checkstack.pl                      Makefile.kasan
checksyscalls.sh                   Makefile.kcov
checkversion.pl                    Makefile.lib
clang-version.sh                   Makefile.modbuiltin
cleanfile                          Makefile.modfinal
cleanpatch                         Makefile.modinst
coccicheck                         Makefile.modpost
coccinelle                         Makefile.modsign
config                             Makefile.package
conmakehash.c                      Makefile.ubsan
const_structs.checkpatch           makelst
decodecode                         markup_oops.pl
decode_stacktrace.sh               mkcompile_h
depmod.sh                          mkmakefile
diffconfig                         mksysmap
documentation-file-ref-check       mkuboot.sh
dtc                                mod
export_report.pl                   module-common.lds
extract-cert.c                     modules-check.sh
extract-ikconfig                   namespace.pl
extract-module-sig.pl              nsdeps
extract-sys-certs.pl               objdiff
extract-vmlinux                    package
extract_xc3028.pl                  parse-maintainers.pl
faddr2line                         patch-kernel
file-size.sh                       pnmtologo.c
find-unused-docs.sh                profile2linkerlist.pl
gcc-goto.sh                        prune-kernel
gcc-ld                             recordmcount.c
gcc-plugins                        recordmcount.h
gcc-plugin.sh                      recordmcount.pl
gcc-version.sh                     selinux
gcc-x86_32-has-stack-protector.sh  setlocalversion
gcc-x86_64-has-stack-protector.sh  show_delta
gdb                                sign-file.c
gen_compile_commands.py            sortextable.c
gen_ksymdeps.sh                    sortextable.h
genksyms                           spdxcheck.py
get_abi.pl                         spdxcheck-test.sh
get_dvb_firmware                   spelling.txt
get_maintainer.pl                  sphinx-pre-install
gfp-translate                      split-man.pl
headerdep.pl                       stackdelta
headers_check.pl                   stackusage
headers_install.sh                 subarch.include
insert-sys-cert.c                  tags.sh
kallsyms.c                         tools-support-relr.sh
Kbuild.include                     tracing
kconfig                            unifdef.c
Kconfig.include                    ver_linux
kernel-doc                         xen-hypercalls.sh
ksymoops                           xz_wrap.sh
```

so looks like this dir gets emptied **after** a kernel is compiled. For me at least. Do you use genkernel? I do.

About unfetchable files, it stops at the very first one:
```
>>> Emerging (1 of 1) net-misc/dahdi-3.1.0::testworld
>>> Downloading 'https://downloads.asterisk.org/pub/telephony/dahdi-linux/releases/dahdi-linux-3.1.0.tar.gz'
--2020-01-06 07:15:53--  https://downloads.asterisk.org/pub/telephony/dahdi-linux/releases/dahdi-linux-3.1.0.tar.gz
Resolving downloads.asterisk.org... 76.164.171.238, 2001:470:e0d4::ee
Connecting to downloads.asterisk.org|76.164.171.238|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1645832 (1.6M) [application/x-gzip]
Saving to: '/var/cache/distfiles/dahdi-linux-3.1.0.tar.gz.__download__'

/var/cache/distfile 100%[===================>]   1.57M  1.15MB/s    in 1.4s    

2020-01-06 07:15:55 (1.15 MB/s) - '/var/cache/distfiles/dahdi-linux-3.1.0.tar.gz.__download__' saved [1645832/1645832]

 * dahdi-linux-3.1.0.tar.gz BLAKE2B SHA512 size ;-) ...                       [ ok ]
>>> Downloading 'https://www.junghanns.net/downloads/jnet-dahdi-drivers-1.0.14.tar.gz'
--2020-01-06 07:15:55--  https://www.junghanns.net/downloads/jnet-dahdi-drivers-1.0.14.tar.gz
Resolving www.junghanns.net... 89.238.81.224
Connecting to www.junghanns.net|89.238.81.224|:443... failed: Connection refused.
!!! Couldn't download 'jnet-dahdi-drivers-1.0.14.tar.gz'. Aborting.
 * Fetch failed for 'net-misc/dahdi-3.1.0', Log file:
 *  '/tmp/portage/net-misc/dahdi-3.1.0/temp/build.log'

>>> Failed to emerge net-misc/dahdi-3.1.0
```
You can test this by defining GENTOO_MIRRORS="" emerge ...

It's not an issue because they are hosted in mirrors, I'm just thinking if upstream has dropped these files (or updated), do they have function here or should they be updated as well?

Reply:
Connecting to www.junghanns.net|89.238.81.224|:443... failed: Connection refused.

So let's hope that's a temporary issue.

Regarding genkernel, no, I don't.  I've got a static .config that we deploy to (almost) all our hosts "as is".  I've tried to use genkernel like 15+ years back, ran into many, many problems.  And just never looked back since.  I generally recommend people to use genkernel unless they know what they're doing.

Thanks for the GENTOO_MIRRORS="" trick.

Regarding your kernel, more likely "linux-5.4_p3-pf" was unmerged, and what's left ar orphan files.  equery belongs should be able to confirm for you (but you already know that I believe).

Reply:
> Regarding your kernel, more likely "linux-5.4_p3-pf" was unmerged, and what's left ar orphan files. equery belongs should be able to confirm for you (but you already know that I believe).

Nope, still have older kernels merged. They have pretty empty scripts/ dir, but the one I built today keeps stuff in it. So I **believe** it's due to genkernel update, but can't confirm. Whatever it may be, it seems "*fixed*" now.

Reply:
Patch added into the patch set external file, so the only update here is the Manifest.